### PR TITLE
Implement delta adsc

### PR DESCRIPTION
**Pull request by hanxiaop**
**Please provide a description of this PR:**
Related to https://github.com/istio/istio/issues/47949.

### CODE REVIEW

File: pkg/adsc/delta_adsc.go [hunk 1]
```diff
@@ -0,0 +1,1174 @@
...
+
+	url string
+
+	watchTime time.Time
+
+	// InitialLoad tracks the time to receive the initial configuration.
+	InitialLoad time.Duration
+
+	// httpListeners contains received listeners with a http_connection_manager filter.
+	httpListeners map[string]*listener.Listener
```
Comment (howardjohn) [inline pkg/adsc/delta_adsc.go:88]: I would prefer we make the new client a better client, not just converted from SotW client. The original adsc was designed as a testing client, and is very poorly suited for production use cases. This is what I worked oln in https://github.com/howardjohn/istio/commit/ba3bd91030b5999aca9391a4f395511eacb7cb58

Comment (hanxiaop) [reply pkg/adsc/delta_adsc.go:88]: Thanks, I've seen https://github.com/istio/istio/issues/36297#issuecomment-1759739439. I'll cherry-pick from what you have and continue with the work.

File: istioctl/pkg/internaldebug/internal-debug.go [hunk 2]
```diff
@@ -29,48 +28,18 @@ import (
 	"istio.io/istio/istioctl/pkg/util"
 	"istio.io/istio/istioctl/pkg/writer/pilot"
 	v3 "istio.io/istio/pilot/pkg/xds/v3"
-	"istio.io/istio/pkg/kube"
 )
 
-func HandlerForRetrieveDebugList(kubeClient kube.CLIClient,
```
Comment (hzxuzhonghu) [inline istioctl/pkg/internaldebug/internal-debug.go:35]: this  seems not related, so do not couple it with delta  adsc

File: pilot/pkg/xds/v3/model.go [hunk 3]
```diff
@@ -35,6 +35,8 @@ const (
 	ProxyConfigType = resource.APITypePrefix + "istio.mesh.v1alpha1.ProxyConfig"
 	// DebugType requests debug info from istio, a secured implementation for istio debug interface.
 	DebugType                 = "istio.io/debug"
+	ConfigDumpType            = envoyTypePrefix + "admin.v3.ConfigDump"
+	ClientConfigType          = envoyTypePrefix + "service.status.v3.ClientConfig"
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/v3/model.go:39]: ditto, unrelated

File: pilot/pkg/xds/discovery.go [hunk 4]
```diff
@@ -560,6 +561,7 @@ func (s *DiscoveryServer) InitGenerators(env *model.Environment, systemNameSpace
 	s.Generators["api/"+TypeURLConnect] = s.StatusGen
 
 	s.Generators["event"] = s.StatusGen
+	s.Generators[resource.APITypePrefix+"envoy.service.status.v3.ClientConfig"] = s.StatusGen
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply pilot/pkg/xds/discovery.go:564]: These are some codes unrelated, I'm trying to cleanup, thanks!

File: pkg/adsc2/client.go [hunk 5]
```diff
@@ -0,0 +1,770 @@
...
+		}
+		h.sub.Insert(key)
+	}
+}
+
+func (h *handlerContext) Reject(reason error) {
+	h.nack = reason
+}
+
+type Config struct {
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply pkg/adsc2/client.go:107]: I'm going to check what adsc has right now and abstract things.

**Comment by hanxiaop** [comment]
@hzxuzhonghu @howardjohn PTAL, thanks

**Review by howardjohn — COMMENTED** [review]
Overall looks good to me. Since I wrote some of this code I will leave to others to approve though

File: istioctl/pkg/internaldebug/internal-debug.go [hunk 6]
```diff
@@ -36,10 +36,10 @@ func HandlerForRetrieveDebugList(kubeClient kube.CLIClient,
 	centralOpts clioptions.CentralControlPlaneOptions,
 	writer io.Writer,
 	istioNamespace string,
-) (map[string]*discovery.DiscoveryResponse, error) {
+) (map[string]*discovery.DeltaDiscoveryResponse, error) {
```
Comment (hzxuzhonghu) [inline istioctl/pkg/internaldebug/internal-debug.go:39]: I do not get the intention here, for debug it is a one time behavior. I cannot see the benefit to change to delta xds

Comment (hanxiaop) [reply istioctl/pkg/internaldebug/internal-debug.go:39]: The intention was to avoid using the stow client, and I think we will eventually need to switch to delta xDS, including the istioctl client usage as well. Currently, there's no benefit for status/debug gen, as they only retrieve resources once. I think it is simpler to make the switch, and right now the only usage of ADSC besides tests are istioctl commands, so I included the change in the same PR.

> I donot think we should transit to using the delta xds in istioctl immediately

Do you think we need to switch tests first and then other things?

File: istioctl/pkg/xds/client.go [hunk 7]
```diff
@@ -41,35 +43,46 @@ const (
 var tokenAudiences = []string{"istio-ca"}
 
 // GetXdsResponse opens a gRPC connection to opts.xds and waits for a single response
-func GetXdsResponse(dr *discovery.DiscoveryRequest, ns string, serviceAccount string, opts clioptions.CentralControlPlaneOptions,
+func GetXdsResponse(dr *discovery.DeltaDiscoveryRequest, ns, serviceAccount string, opts clioptions.CentralControlPlaneOptions,
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply istioctl/pkg/xds/client.go:46]: Replied above.

**Comment by hanxiaop** [comment]
> Would better to have a document about how `resourceCache` and `tree` collaborate to make it a stateful client.

Will do.

File: pkg/adsc2/client.go [hunk 8]
```diff
@@ -0,0 +1,620 @@
...
+	deltaXDSUpdates chan *discovery.DeltaDiscoveryResponse
+
+	resourceCache sync.Map
+
+	errChan chan error
+}
+
+func (c *Client) trigger(ctx *handlerContext, url string, r *discovery.Resource, event Event) error {
+	var res proto.Message
+	if isDebugType(url) {
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply pkg/adsc2/client.go:132]: Pasted the wrong link?

File: pkg/adsc2/client.go [hunk 9]
```diff
@@ -0,0 +1,620 @@
...
+
+	// sendNodeMeta is set to true if the connection is new - and we need to send node meta
+	sendNodeMeta atomic.Bool
+
+	mutex sync.RWMutex
+	// Last received message, by type
+	received        map[string]*discovery.DeltaDiscoveryResponse
+	deltaXDSUpdates chan *discovery.DeltaDiscoveryResponse
+
+	resourceCache sync.Map
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply pkg/adsc2/client.go:125]: We store the value to manage the resource when it is removed; otherwise, from the DeltaDiscoveryResponse, we would only know the name of the removed resource. Currently, we watch specific types/resources instead of all resources to minimize memory usage. Would it be better to make the cache optional for cases where tracking deleted resources is not necessary?

**Comment by hanxiaop** [comment]
Accidentally closed by istio bot.

File: istioctl/pkg/internaldebug/internal-debug.go (hunk 6 above)
Comment (hzxuzhonghu) [reply istioctl/pkg/internaldebug/internal-debug.go:39]: To be clear, i donot think we will remove supporting Stow. Delta is an option with feature flag. 

And implementing adsc delta is not equal to transiting to delta xds.

File: pkg/adsc2/client.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pkg/adsc2/client.go:132]: I mean `isDebugType` is checked by the caller already

File: pkg/adsc2/client.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pkg/adsc2/client.go:125]: >otherwise, from the DeltaDiscoveryResponse, we would only know the name of the removed resource.

I am not sure isn't this satisfying

File: istioctl/pkg/internaldebug/internal-debug.go (hunk 6 above)
Comment (hanxiaop) [reply istioctl/pkg/internaldebug/internal-debug.go:39]: I see, thanks for the clarification! I'll remove the command changes, and if we need them later, we can do a cherry-pick. This should make the PR cleaner.

File: pkg/adsc2/client.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (hanxiaop) [reply pkg/adsc2/client.go]: I've adde an option, and also some explainations.

https://github.com/istio/istio/blob/1dd980cab5451814af3442e4fee6a413efd89636/pkg/adsc/delta.go#L96-L129

 Can this address your concern?

**Comment by hzxuzhonghu** [comment]
will take a look later

File: pkg/adsc/delta.go [hunk 10]
```diff
@@ -0,0 +1,703 @@
...
+			Children: make(keySet),
+		}
+		c.tree[key] = parentNode
+	}
+
+	wildcardKey := resourceKey{TypeURL: key.TypeURL}
+	wildNode, wildFound := c.tree[wildcardKey]
+	if wildFound {
+		wildNode.Children.Insert(key)
+		parentNode.Parents.Insert(wildcardKey)
```
Comment (hzxuzhonghu) [inline pkg/adsc/delta.go:466]: can you add more documents about it

File: pkg/adsc/delta.go [hunk 11]
```diff
@@ -0,0 +1,703 @@
...
+//  3. State Synchronization: Post-processing the delta response, the client updates its internal state. This involves:
+//     - Acknowledgements and Errors: Communicating acknowledgements or errors back to the server based on the
+//     processing outcome. In cases of error or rejection, a Nack can be sent using HandlerContext.Reject.
+//     - Dependency Updates: Triggering requests for dependent resources. These dependencies are established via
+//     HandlerContext.RegisterDependency.
+type Client struct {
+	cfg      *DeltaADSConfig
+	handlers map[string]HandlerFunc
+	// Map of resources key to its node
+	tree map[resourceKey]resourceNode
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply pkg/adsc/delta.go:196]: > I am not sure what is a resource's parent and children.

And here's an example of the resource node:

https://github.com/istio/istio/blob/2264ec09b4b69ccb745bc5901add0fc4c5ff1da4/pkg/adsc/delta.go#L57-L87

File: pkg/adsc/adsc_test.go [hunk 12]
```diff
@@ -406,7 +407,11 @@ func TestADSC_handleMCP(t *testing.T) {
 	adsc := &ADSC{
 		VersionInfo: map[string]string{},
 		Store:       memory.Make(collections.Pilot),
-		cfg:         &Config{Revision: rev},
+		cfg: &ADSConfig{
+			Config: Config{
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply pkg/adsc/adsc_test.go:411]: I think it's not working because it's not a pointer.

File: pkg/adsc/delta.go [hunk 13]
```diff
@@ -0,0 +1,703 @@
...
+	// sendNodeMeta is set to true if the connection is new - and we need to send node meta
+	sendNodeMeta atomic.Bool
+
+	mutex sync.RWMutex
+	// Last received message, by type
+	received        map[string]*discovery.DeltaDiscoveryResponse
+	deltaXDSUpdates chan *discovery.DeltaDiscoveryResponse
+
+	// resourceCache is used to cache the resource received from the delta stream,
+	// this is only used for triggering the delete event when the resource is removed from the stream.
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply pkg/adsc/delta.go:213]: After thinking I think this is right, we may not need the cache as having the name is sufficient to trigger the deletion. I'll revise.

File: pkg/adsc/delta.go [hunk 14]
```diff
@@ -0,0 +1,703 @@
...
+		}
+	}
+}
+
+func (c *Client) handleDeltaResponse(d *discovery.DeltaDiscoveryResponse) error {
+	var rejects []error
+	allAdds := map[string]set.Set[string]{}
+	allRemoves := map[string]set.Set[string]{}
+	ctx := &handlerContext{}
+	defer func() {
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply pkg/adsc/delta.go]: For debugType, we need to record and use the wait function to wait for resources. Otherwise, we won't know when the handling of the subscribed resources has ended.

**Comment by hanxiaop** [comment]
@hzxuzhonghu Could you take a look again? Thanks!

**Comment by hzxuzhonghu** [comment]
Yeah, busy these days, will try to understand the `tree`

File: pkg/adsc/delta.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pkg/adsc/delta.go:190]: Thanks, it is clear.

File: pilot/test/xdstest/extract.go [hunk 15]
```diff
@@ -68,6 +69,49 @@ func ExtractRoutesFromListeners(ll []*listener.Listener) []string {
 	return routes
 }
 
+func ExtractClusterSecretResources(t test.Failer, c *cluster.Cluster) []string {
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/test/xdstest/extract.go:72]: Oh, ic it is for dependency register

File: pilot/test/xdstest/extract.go [hunk 16]
```diff
@@ -68,6 +69,49 @@ func ExtractRoutesFromListeners(ll []*listener.Listener) []string {
...
+	resourceNames := sets.New[string]()
+	var sockets []*core.TransportSocket
+	if c.TransportSocket != nil {
+		sockets = append(sockets, c.TransportSocket)
+	}
+	for _, ts := range c.TransportSocketMatches {
+		sockets = append(sockets, ts.TransportSocket)
```
Comment (hzxuzhonghu) [inline pilot/test/xdstest/extract.go:79]: nit: for proxy protocol, this is not very correct. But since only used test. Doesnot matter much

File: pkg/adsc/delta.go [hunk 17]
```diff
@@ -0,0 +1,739 @@
...
+
+	// initialWatches is the list of resources we are watching on startup
+	initialWatches []resourceKey
+
+	// sendNodeMeta is set to true if the connection is new - and we need to send node meta
+	sendNodeMeta atomic.Bool
+
+	mutex sync.RWMutex
+	// Last received message, by type
+	received        map[string]*discovery.DeltaDiscoveryResponse
```
Comment (hzxuzhonghu) [inline pkg/adsc/delta.go:203]: What's this for?

IIRC, it is used for async response. The it is not quite right, because we cannot gurantee the orders.

Comment (hzxuzhonghu) [inline pkg/adsc/delta.go:203]: I would also rename it `lastReceived`

File: pkg/adsc/delta.go [hunk 18]
```diff
@@ -0,0 +1,739 @@
...
+			resourceName = ""
+		}
+		key := resourceKey{
+			Name:    resourceName,
+			TypeURL: typeURL,
+		}
+		existing, f := c.tree[key]
+		if f {
+			// We are watching directly now, so erase any parents
+			existing.Parents = nil
```
Comment (hzxuzhonghu) [inline pkg/adsc/delta.go:346]: should we erase child as well, initWatch is called after a new connect or reconnect. Seems a totally new initialization

File: pkg/adsc/delta.go [hunk 19]
```diff
@@ -0,0 +1,739 @@
...
+			}
+		}
+		c.initialWatches = append(c.initialWatches, key)
+	}
+}
+
+func (c *Client) handleRecv() {
+	// We connected, so reset the backoff
+	if c.cfg.BackoffPolicy != nil {
+		c.cfg.BackoffPolicy.Reset()
```
Comment (hzxuzhonghu) [inline pkg/adsc/delta.go:360]: why donot reset in `Run` after success

File: pkg/adsc/delta.go [hunk 20]
```diff
@@ -0,0 +1,739 @@
...
+			default:
+			}
+			// if 'reconnect' enabled - schedule a new Run
+			if c.cfg.BackoffPolicy != nil {
+				time.AfterFunc(c.cfg.BackoffPolicy.NextBackOff(), c.reconnect)
+			} else {
+				c.Close()
+				c.WaitClear()
+				c.deltaXDSUpdates <- nil
+				close(c.errChan)
```
Comment (hzxuzhonghu) [inline pkg/adsc/delta.go:378]: move these cleanup to `Close`

File: pkg/adsc/delta.go [hunk 21]
```diff
@@ -0,0 +1,739 @@
...
+			TypeURL: r.Resource.TypeUrl,
+		}
+		c.establishResource(parentKey)
+		if ctx.nack != nil {
+			rejects = append(rejects, ctx.nack)
+			// On NACK, do not apply resource changes
+			continue
+		}
+
+		remove, add := c.tree[parentKey].Children.Diff(ctx.sub)
```
Comment (hzxuzhonghu) [inline pkg/adsc/delta.go:424]: Confused about the `remove`.  DeltaResponse has a removed field, why do you need to compare local cache with the response to get the removed resources

Comment (hzxuzhonghu) [reply pkg/adsc/delta.go:424]: NVM, got it.

File: pkg/adsc/delta.go [hunk 22]
```diff
@@ -0,0 +1,739 @@
...
+	if !f && !wildFound {
+		// We are receiving an unwanted resource, silently ignore it.
+		deltaLog.Debugf("Received unsubscribed resource: %v, %v", key, c.tree)
+	}
+}
+
+func (c *Client) relate(parent, child resourceKey) {
+	parentNode, f := c.tree[parent]
+	if !f {
+		deltaLog.Fatalf("Failed to relate resource: unknown parent: %v, %v", parent, c.tree)
```
Comment (hzxuzhonghu) [inline pkg/adsc/delta.go:514]: mark as error?

File: pkg/adsc/delta.go (hunk 17 above)
Comment (hanxiaop) [reply pkg/adsc/delta.go]: > I would also rename it lastReceived

I think naming this "lastReceived" makes sense? As you said, this doesn't guarantee the order, so users will need to implement custom checks on what was received. Like:

https://github.com/istio/istio/blob/06ad655977d4619343f530280d0afeece9d95945/pkg/adsc/delta_test.go#L437-L454


However, if we don't include this logic, it would require users to implement more complex logic to determine when the response for the watched resources has terminated.

File: pilot/test/xdstest/extract.go (hunk 16 above)
Comment (hanxiaop) [reply pilot/test/xdstest/extract.go:79]: Haven't investigated this yet, can we update it when we need?

File: pkg/adsc/delta.go (hunk 18 above)
Comment (hanxiaop) [reply pkg/adsc/delta.go:358]: I think this is reasonable, the relationship will be rebuilt after processing responses.

File: pkg/adsc/delta.go (hunk 19 above)
Comment (hanxiaop) [reply pkg/adsc/delta.go]: Sounds good, thanks

File: pkg/adsc/delta.go (hunk 22 above)
Comment (hanxiaop) [reply pkg/adsc/delta.go:518]: I think this actually should not happen; perhaps error may not be enough?

**Review by hzxuzhonghu — APPROVED** [review]
Thanks @hanxiaop for so many rounds of fix

**Comment by hanxiaop** [comment]
> Thanks @hanxiaop for so many rounds of fix

Thanks for the careful review!

**Review by GregHanson — APPROVED** [review]
LGTM

**Comment by hanxiaop** [comment]
@istio/wg-test-and-release-maintainers PTAL, thanks!

**Comment by howardjohn** [comment]
I'll take a look tomorrow

**Review by howardjohn — APPROVED** [review]

**Comment by YonkaFang** [comment]
I see that the `delta-adsc` is currently not being used, and there is also no support for MCP resources. What is the current progress and the upcoming plan for this?

File: pkg/adsc/delta.go [hunk 23]
```diff
@@ -0,0 +1,736 @@
...
+func newProto(tt string) proto.Message {
+	name := protoreflect.FullName(strings.TrimPrefix(tt, resource.APITypePrefix))
+	t, err := protoregistry.GlobalTypes.FindMessageByName(name)
+	if err != nil || t == nil {
+		return nil
+	}
+	return t.New().Interface()
+}
+
+func (c *Client) Run(ctx context.Context) error {
```
Comment (wulianglongrd) [inline pkg/adsc/delta.go:235]: Hello, is this feature currently enabled? I don't see this method called anywhere.

File: pilot/pkg/bootstrap/configcontroller.go [hunk 24]
```diff
@@ -232,23 +232,25 @@ func (s *Server) initConfigSources(args *PilotArgs) (err error) {
...
-			xdsMCP, err := adsc.New(srcAddress.Host, &adsc.Config{
-				Namespace: args.Namespace,
-				Workload:  args.PodName,
-				Revision:  args.Revision,
-				Meta: model.NodeMetadata{
-					Generator: "api",
-					// To reduce transported data if upstream server supports. Especially for custom servers.
-					IstioRevision: args.Revision,
-				}.ToStruct(),
+			xdsMCP, err := adsc.New(srcAddress.Host, &adsc.ADSConfig{
```
Comment (wulianglongrd) [inline pilot/pkg/bootstrap/configcontroller.go:235]: It looks like you need to be able to switch to `adsc.NewDelta` here.

Comment (hanxiaop) [reply pilot/pkg/bootstrap/configcontroller.go:235]: Currently it has not been used in tests or code yet, but it will be.

Comment (wulianglongrd) [reply pilot/pkg/bootstrap/configcontroller.go:235]: Thanks, got it.
